jax/_src/scipy/special.py

```diff
@@ -2582,3 +2584,78 @@ def hyp1f1(a: ArrayLike, b: ArrayLike, x: ArrayLike) -> Array:
   lambda b_dot, primal_out, a, b, x: _hyp1f1_b_derivative(a, b, x) * b_dot,
   lambda x_dot, primal_out, a, b, x: _hyp1f1_x_derivative(a, b, x) * x_dot
 )
+
+
+# This interface is designed to be compatible with the future Array API
+# interface, which has not been specified yet as of June 2024.  We immitate its
+# sum function's use of keyword-only and positional-only parameters.
+def softmax(x: ArrayLike,
+            /,
+            *,
+            axis: int | tuple[int, ...] | None = None,
+            ) -> Array:
+  r"""Softmax function.
+
+  Computes the function which rescales elements to the range :math:`[0, 1]`
+  such that the elements along :code:`axis` sum to :math:`1`.
+
+  .. math ::
+    \mathrm{softmax}(x) = \frac{\exp(x_i)}{\sum_j \exp(x_j)}
+
+  Args:
+    x : input array
+    axis: the axis or axes along which the softmax should be computed. The
+      softmax output summed across these dimensions should sum to :math:`1`.
+
+  Returns:
+    An array.
+
+  Note:
+    If any input values are ``+inf``, the result will be all ``NaN``: this
+    reflects the fact that ``inf / inf`` is not well-defined in the context of
+    floating-point math.
+
+  See also:
+    :func:`log_softmax`
+  """
+  if isinstance(axis, int) and axis < 0:
+    raise ValueError
```

Comment:
Supporting -1 seems like it will be a giant bug magnet since [the Array API treats -1 as the last axis](https://data-apis.org/array-api/latest/API_specification/indexing.html#single-axis-indexing).    Why not just have users use `None`?

Reply:
We're not implementing the array API here, we're implementing the `scipy.special` API.

Reply:
And my understanding is that `axis=None` is equivalent to `axis=tuple(range(x.ndim))`, so `axis=-1` is quite different than `axis=None`.

Reply:
Okay, you're right.